src/darwin/Framework/CHIP/MTREndpointInfo.mm

```diff
@@ -0,0 +1,210 @@
+                         partsList:(NSArray<NSNumber *> *)partsList
+{
+    self = [super init];
+    _endpointID = endpointID.unsignedShortValue;
+    _deviceTypes = [deviceTypes copy];
+    _partsList = [partsList copy];
+    _children = @[];
+    return self;
+}
+
+- (NSNumber *)endpointID
+{
+    return @(_endpointID);
+}
+
+- (NSString *)description
+{
+    return [NSString stringWithFormat:@"<%@ %u>", self.class, _endpointID];
+}
+
++ (BOOL)populateChildrenForEndpoints:(NSDictionary<NSNumber *, MTREndpointInfo *> *)endpoints
+{
+    // Populate the child list of each endpoint, ensuring no cycles (these are disallowed
+    // by the spec, but we need to ensure we don't create a retain cycle even under invalid
+    // input). Conservatively assume all endpoints use the Full-Family Pattern.
+    // Refer to "Endpoint Composition" in the Matter specification for details.
+    MTREndpointInfo * root = endpoints[@0];
+    if (root == nil) {
+        MTR_LOG_ERROR("Missing root endpoint, not populating endpoint hierarchy");
+        return NO;
+    }
+
+    // Perform a depth-first search with an explicit stack and create a list of endpoint
+    // IDs in reverse topological order. The _parentID field is used mark traversal states.
+    BOOL valid = YES;
+    std::deque<EndpointId> deque; // stack followed by sorted list
+    deque.emplace_front(root->_endpointID);
+    for (;;) {
+        EndpointId endpointID = deque.front();
+        MTREndpointInfo * endpoint = endpoints[@(endpointID)];
+        if (endpoint->_mark == EndpointMark::NotVisited) {
+            endpoint->_mark = EndpointMark::Visiting;
+            for (NSNumber * partNumber in endpoint->_partsList) {
+                MTREndpointInfo * descendant = endpoints[partNumber];
+                if (!descendant) {
+                    MTR_LOG_ERROR("Warning: PartsList of endpoint %u references non-existant endpoint %u",
+                        endpointID, partNumber.unsignedShortValue);
+                    valid = NO;
+                } else if (descendant->_mark == EndpointMark::NotVisited) {
+                    deque.emplace_front(descendant->_endpointID);
+                } else if (descendant->_mark == EndpointMark::Visiting) {
+                    MTR_LOG_ERROR("Warning: Cyclic endpoint composition involving endpoints %u and %u",
+                        descendant->_endpointID, endpointID);
+                    valid = NO;
+                }
+            }
+        } else if (endpoint->_mark == EndpointMark::Visiting) {
+            endpoint->_mark = EndpointMark::Visited;
+            deque.pop_front(); // remove from stack
+            deque.emplace_back(endpointID); // add to sorted list
+            if (endpointID == root->_endpointID) {
+                break; // visited the root, DFS traversal done
+            }
+        } else /* endpoint->_mark == EndpointMark::Visited */ {
+            deque.pop_front(); // nothing else to do
```

Comment:
So when would we reach this mark == Visited case?  Seems to me that it's not reachable, since we put things on the end when we mark them visited, break out when we do that with the root endpoint, and start off with the root endpoint.....  What am I missing?

Reply:
I've added a comment there and a couple more tests. This can actually happen in two ways:

In `testPopulateChildren2`, endpoint 2 is visited due to being a child of 1 (and the visit completed), then we visit it again from the root node. This is a valid configuration, the same as `testPopulateChildren` simply with a different PartsList order.

In `testPopulateChildrenInvalidNonTree`, endpoint 6 is a child of both 1 and 5, and these are in sibling branches of the tree. This is an invalid configuration, but not easily distinguishable from the other scenario.

(The valid / invalid return value is at this point primarily used for unit testing, since in practice I think we should be tolerant and not let some errant endpoint with invalid meta-data deter us from interacting with an otherwise valid device.)